### ISSUE TITLE
[HOLD for payment 2023-01-19] [$2000] Feature Request: Open link in Desktop app if it is installed and not in browser

### ISSUE THREAD

**Issue by MonilBhavsar** [Expensify/App#10893]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Have desktop app installed
2. Create a new account on desktop app
3. Click the validation link from email

## Expected Result:

Link should open in Desktop app

## Actual Result:

Link always open in browser

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?
For users:
They can still use the app on web and then move to desktop

For contributors: 
None
I think contributors are not able to test the workflow on the desktop app

## Platform:

Where is this issue occurring?

- Web
- Desktop

**Version Number:** 
**Reproducible in staging?:** 
**Reproducible in production?:** 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Upwork job URL:** https://www.upwork.com/jobs/~01c5b3dd89b6a749b1
**Issue reported by:** @MonilBhavsar 
**Slack conversation:**  https://expensify.slack.com/archives/C01GTK53T8Q/p1662630927711119

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

Slack handles this nicely! If desktop app is installed, Link opens in app and they ask if one wants to open it in the browser.

<img width="1792" alt="Screenshot 2022-09-08 at 11 20 17 AM" src="https://user-images.githubusercontent.com/32012005/189099067-c14bacb9-847b-46b8-a743-b05a860c655f.png">

**Comment by madmax330** [comment]
@MonilBhavsar I think this can be external yeah?

**Comment by MonilBhavsar** [comment]
Yes 👍

**Comment by madmax330** [comment]
Cool

**Comment by michaelhaxhiu** [comment]
Job is posted to upwork - https://www.upwork.com/jobs/~01c5b3dd89b6a749b1

**Comment by luacmartins** [comment]
FYI this issue would also happen coming from the android or iOS apps. So if we go with this solution we should also deeplink it in those apps.

**Comment by luacmartins** [comment]
Nvm my comment above, that link is already deeplinked in the native apps

**Comment by NikkiWines** [comment]
Not overdue, waiting on proposals.

**Comment by michaelhaxhiu** [comment]
doubling

**Comment by michaelhaxhiu** [comment]
Doubled

**Comment by michaelhaxhiu** [comment]
Waiting on Santhosh to review the very detailed proposal above

**Comment by michaelhaxhiu** [comment]
Wait a second... maybe we shouldn't double actually, since we are laser-focused on delegating **bugs** to contributors right now. 

Let's pause on doubling. @b1tjoy's proposal is still valid for consideration though.

**Comment by NikkiWines** [comment]
> I think we should have an open in desktop popup first before loading the page password page on the web, on confirmation we should open it on the desktop.

Agreed on this, we shouldn't show the same view in both web + desktop. Maybe we could also have some sort of interstitial page like what Slack has, but that could be a follow up improvement. 

<img width="538" alt="Screen Shot 2022-10-14 at 16 05 38" src="https://user-images.githubusercontent.com/16747822/195866823-c65d5d6e-ce0a-4b79-8858-ab71d3f84fde.png">

**Comment by Santhosh-Sellavel** [comment]
The feature should not be specific to the Set Password Page, we need a comprehensive solution for the whole app or at least links configured [here](https://github.com/Expensify/App/blob/main/.well-known/apple-app-site-association)

**Comment by NikkiWines** [comment]
@Santhosh-Sellavel good point, we'd need a solution that works for any link that a user would open in a browser

**Comment by michaelhaxhiu** [comment]
@b1tjoy are you interested in adapting your proposal to accommodate the latest feedback? Doubled the job value to encourage that :)

**Comment by cowboycito** [comment]
Simple and straightforward proposal since I don't have much time available at this exact moment, but could weigh in a little more later today.

# Proposal
Create a new `DeeplinkWrapper` component that would be a global wrapper of the web app and would listen to URL changes. At first we could decide on a subset of routes we want to have deeplink (we could use [this](https://github.com/Expensify/App/blob/main/.well-known/apple-app-site-association) as a start). The `DeeplinkWrapper` component would match the current route with the routes we want to deeplink and, if it's a match, we would display the popup with the option for the user to open that specific page on the Mac desktop app.
If the user do not want to do that, I suggest we use some simple storage service (like `localStorage`) to store that the user did NOT want to deeplink that specific page and also the timestamp, which we would then take into consideration the next time the user accesses that page, so we don't keep bothering them with the deeplink popup every single time (in a timespan) they open that route.
Besides that a simple change would be necessary on the desktop app to handle these deeplink URLs.

Later today I can do some small coding proposal here on the comments if you all think this is a good solution.

**Comment by cowboycito** [comment]
# Proposal

Ok so here's a more formal proposal for the issue.

A new component called `DeeplinkWrapper` would be created as below (only for web - and desktop -, on mobile it would just return `children`):

```
import  PropTypes  from  'prop-types';
import {PureComponent} from  'react';
import {deeplinkRoutes} from  './deeplinkRoutes';

const  propTypes  = {
	/** Children to render. */
	children: PropTypes.node.isRequired,
};

class  DeeplinkWrapper  extends  PureComponent {
	constructor(props) {
		super(props);

		// check if pathname matches with deeplink routes
		const  pathname  = window.location.pathname
		const  matchedRoute  = deeplinkRoutes.find(route => {
			const  routeRegex  =  new  RegExp(route.pattern)
			return routeRegex.test(pathname)
		})

		if (!!matchedRoute &&  this.isMacOSWeb()) {) {
			this.state  = {
				deeplinkMatch:  true,
			};

			// open desktop app
		}
		else {
			deeplinkMatch:  false,
		}
	}

	isMacOSWeb() {
		if (
			typeof navigator ===  'object'  &&
			typeof navigator.userAgent  ===  'string'  &&
			/Mac/i.test(navigator.userAgent) &&
			!/Electron/i.test(navigator.userAgent)
		) {
			return  true;
		}

		return false;
	}

	render() {
		if (this.state.matchedDeeplinkRoute) {
			// return Slack-like page with
			// option for the user to open
			// page on the browser
		}

		return  this.props.children;
	}
}

DeeplinkWrapper.propTypes  = propTypes;
export  default  DeeplinkWrapper;
```

Please note that `isMacOSWeb` is a courtesy from @b1tjoy

Besides that we would just need a simple way of handling the deeplink URL on the dekstop app, and @b1tjoy seems like it'll do the trick.

This proposal is a bit different from the one I suggested in the other comment, since I figured it would be very annoying for the user if we kept asking to open the desktop app every single time they navigate to a subroute of a route we want to deeplink (e.g.: any sobroute in `/bank-account/*`) .

**Comment by Santhosh-Sellavel** [comment]
@cowboycito 
Thanks for the proposal

> Besides that we would just need a simple way of handling the deeplink URL on the dekstop app, and @b1tjoy's solution for that seems like it'll do the trick.

Like you used `isMacOSWeb`  you could've added that also in your proposal.

How we will be using `DeeplinkWrapper` component?

Can you break it done it to steps like [here](https://github.com/Expensify/App/issues/10893#issuecomment-1272495980) send an updated proposal in a new comment. thanks!

**Comment by cowboycito** [comment]
# Proposal

A new component called `DeeplinkWrapper` would be created as below (only for web - on mobile and desktop it would just return `children`):

```javascript
import PropTypes from 'prop-types';
import {PureComponent} from 'react';
import {deeplinkRoutes} from './deeplinkRoutes';
import CONST from '../../CONST';
import CONFIG from '../../CONFIG';

const propTypes = {
  /** Children to render. */
  children: PropTypes.node.isRequired,
};

class DeeplinkWrapper extends PureComponent {
  constructor(props) {
    super(props);

    // check if pathname matches with deeplink routes
    const pathname = window.location.pathname;
    const matchedRoute = deeplinkRoutes.find((route) => {
      const routeRegex = new RegExp(route.pattern);
      return routeRegex.test(pathname);
    });

    if (!!matchedRoute && this.isMacOSWeb()) {
      this.state = {
        deeplinkMatch: true,
      };

      const expensifyUrl = new URL(CONFIG.EXPENSIFY.NEW_EXPENSIFY_URL);
	  window.location = `${CONST.DEEPLINK_BASE_URL}${expensifyUrl.host}${pathname}`;
    } else {
      this.state = {
        deeplinkMatch: false,
      };
    }
  }

  isMacOSWeb() {
    if (
      typeof navigator === 'object' &&
      typeof navigator.userAgent === 'string' &&
      /Mac/i.test(navigator.userAgent) &&
      !/Electron/i.test(navigator.userAgent)
    ) {
      return true;
    }

    return false;
  }

  render() {
    if (this.state.matchedDeeplinkRoute) {
      // return Slack-like page with
      // option for the user to open
      // page on the browser
    }

    return this.props.children;
  }
}

DeeplinkWrapper.propTypes = propTypes;
export default DeeplinkWrapper;
```

Please note that `isMacOSWeb` is a courtesy from @b1tjoy.

This component would be used in `src/Expensify.js` as follows:

```diff
render() {
	// Display a blank page until the onyx migration completes
	if (!this.state.isOnyxMigrated) {
		return  null;
	}

	return (
-		<>
+		<DeeplinkWrapper>
			{!this.state.isSplashShown  && (
				<>
					<GrowlNotification  ref={Growl.growlRef} />
					...
				</>
			)}
			
			<NavigationRoot
				onReady={this.setNavigationReady}
				authenticated={this.isAuthenticated()}
			/>
-		</>
+		</DeeplinkWrapper>
	);
}
```

For handling the deeplink URL in the desktop app we would need 2 steps:

1. Configure deeplink scheme in `config/electronBuilder.config.js`
```diff
module.exports  = {
	...
	directories: {
		app:  'desktop',
		output:  'desktop-build',
	},
+	protocols: {
+		name:  'New Expensify',
+		schemes: [CONST.DEEPLINK_BASE_URL],
	},
};
```

2. Configure the main process to handle the deeplink URL (treating both cases where app might be already opened or closed) in `desktop.main.js`
```diff
...
const mainWindow = () => {
+	let deeplinkUrl;
+	let browserWindow;

	const loadURL = __DEV__
		? (win) => win.loadURL(`http://localhost:${port}`)
		: serve({directory: `${__dirname}/www`});

	// Prod and staging set the icon in the electron-builder config, so only update it here for dev
	if (__DEV__) {
		console.debug('CONFIG: ', CONFIG);
		app.dock.setIcon(`${__dirname}/../icon-dev.png`);
		app.setName('New Expensify');
	}

+	app.on('will-finish-launching', () => {
+		app.on('open-url', (event, url) => {
+	    	event.preventDefault();
+			const urlObj = new URL(url);
+			deeplinkURL = `${APP_DOMAIN}${urlObj.pathname}`;
+			if (browserWindow) {
+				browserWindow.loadURL(deeplinkURL);
+			}
+		});
+	});

	return (
		app
			.whenReady()
			.then(() => {
-  				const browserWindow = new BrowserWindow({
+				browserWindow = new BrowserWindow({
          			backgroundColor: '#FAFAFA',
					width: 1200,
					height: 900,
					
				...


	// After initializing and configuring the browser window, load the compiled JavaScript
	.then((browserWindow) => {
-		loadURL(browserWindow);
+		loadURL(browserWindow).then(() => {
+			if (deeplinkUrl) {
+				browserWindow.loadURL(deeplinkUrl);
+			}
+		})
		return browserWindow;
	})
```

And another detail of the "Slack-like" approach is that ideally we would need a design for the page we would show on the browser when the deeplink route is a match, like the image posted by @NikkiWines in the comments above.

@Santhosh-Sellavel this is a more detailed proposal. If you have any questions please let me know.

**Comment by Santhosh-Sellavel** [comment]
Perfect @cowboycito this looks good & well written. I'll review and let you know Thanks!

**Comment by michaelhaxhiu** [comment]
Seems we are evaluating the proposal provided and will have a 👍 or 👎 provided soon.

**Comment by michaelhaxhiu** [comment]
Still waiting on @Santhosh-Sellavel to review the latest proposal, but might be a little delayed due to local holidays.

**Comment by Santhosh-Sellavel** [comment]
Will review it this week, mostly in a couple of days!

**Comment by Santhosh-Sellavel** [comment]
Thanks for the patience everyone!

@cowboycito Looks good, what does `deeplinkRoutes` contain?

**Comment by cowboycito** [comment]
@Santhosh-Sellavel 

```javascript
export const deeplinkRoutes = [
  {
    pattern: /\/r\/*/,
    comment: 'Reports',
  },
  {
    pattern: /\/settings\/*/,
    comment: 'Profile and app settings',
  },
  {
    pattern: /\/setpassword\/*/,
    comment: 'Passoword setup',
  },
  {
    pattern: /\/details\/*/,
    comment: 'Details of another users',
  },
  {
    pattern: /\/v\/*/
    comment: 'Account validation',
  },
  {
    pattern: /\/bank-account\/*/
    comment: 'Bank account setup and its steps',
  },
  {
    pattern: /\/iou\/*/,
    comment: 'I Owe You reports',
  },
  {
    pattern: /\/enable-payments\/*/,
    comment: 'Payments setup',
  },
  {
    pattern: /\/partners\/plaid\/oauth_ios/,
    pattern: '/partners/plaid/oauth_ios',
    comment: 'Plaid setup',
  },
  {
    pattern: /\/statements\/*/,
    comment: 'Wallet statements',
  },
  {
    pattern: /\/concierge\/*/,
    comment: 'Concierge',
  },
];
```

**Comment by Santhosh-Sellavel** [comment]
@NikkiWines So far @cowboycito proposal looks good, let's loop its design and get the page intermediate page designed!

**Comment by Santhosh-Sellavel** [comment]
@cowboycito One more question can you provide a basic skeleton component, with two buttons? 
Open on desktop - opens the link in the app
Continues - continue on the web!

So let's test the POC!

**Comment by cowboycito** [comment]
@Santhosh-Sellavel honestly I feel like this is something that will be done in the PR once my proposal gets accepted. I already dedicated a lot of time doing the proposal itself and answering to questions and I don't plan to spend more time on this until I get the job. Hope that's ok.

**Comment by Santhosh-Sellavel** [comment]
@cowboycito Sorry,  we would definitely need that code to test the proof of concept here since this is a feature request. Without that, we don't have enough to make a hiring decision. You are the preferred candidate here now, please try to give us an updated proposal, thanks!

**Comment by cowboycito** [comment]
@Santhosh-Sellavel ok, I'll try to get it done by Wednesday night and will let you know. I'm fairly new to Expensify's bounty program so I'd like to know if there's any chance we could increase the price of this issue?

**Comment by Santhosh-Sellavel** [comment]
Again this is too early, we can discuss it once the issue is completed.

**Comment by cowboycito** [comment]
@Santhosh-Sellavel no worries, just wanted to know if it was possible

**Comment by cowboycito** [comment]
@Santhosh-Sellavel please take a look at the [iss-10893/desktop-deeplink](https://github.com/cowboycito/App/tree/iss-10893/desktop-deeplink) branch. The POC code is there. Let me know if you have any questions or ideas for improvements.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @cowboycito!

**Comment by michaelhaxhiu** [comment]
@cowboycito we'll aim to give a confirmation on next steps in the next 1-2 days. I'm sure @Santhosh-Sellavel is already moving on it :) And then @NikkiWines can give the approval and we move forward.

**Comment by cowboycito** [comment]
Just a quick note. I literally just noticed that the solution in the POC branch is trying to open the app when you open it on iOS web. That would for sure be fixed in the final solution, just need to investigate the issue a little bit more.

**Comment by Santhosh-Sellavel** [comment]
@cowboycito 
I'm getting a pop-up  without clicking the button, while opening the link
<img width="1187" alt="Screenshot 2022-11-13 at 11 15 28 PM" src="https://user-images.githubusercontent.com/85645967/201536353-7c2807c9-4426-4543-99ef-755dfaa1a952.png">

Clicking the Open in App it's not opened in the desktop app, how to test this?

**Comment by cowboycito** [comment]
@michaelhaxhiu sorry I didn't notice that @Santhosh-Sellavel had posted a message here.
So about statement 1 that's exactly how Slack and Zoom do, and there's no other way to check if the desktop app is installed on the user's computer, so I'm afraid we have no other way to do that.
And about statement 2, I'll do some digging tonight to check if I'm able to replicate the issue and will post an update in the comments here. @Santhosh-Sellavel if you click in the "Open Electron" button in the pop-up, does It open the desktop app correctly?

**Comment by cowboycito** [comment]
@Santhosh-Sellavel I just re-ran the code and everything seems to be working fine for me. Can you give any any specific steps to reproduce the issue that you're having?
Here's a screen recording to demo how things are working right now.


https://user-images.githubusercontent.com/47456024/202316925-8dfb6489-dbbd-40dd-909b-5e15163cc3d8.mp4

**Comment by NikkiWines** [comment]
I agree with @cowboycito, other apps like Slack and Spotify display a similar pop-up if you haven't selected "always allow" 

<img width="738" alt="image" src="https://user-images.githubusercontent.com/16747822/202608408-6baaa025-f92c-4d26-a2e3-9a805a36c2e8.png">

Some basic testing on the POC branch seems to work well so far, though I'd expect the pop-up to say "Open New Expensify" not "Open Electron"

**Comment by cowboycito** [comment]
@NikkiWines I believe that's because we're running the electron app instead of a built version of it. Once my proposal gets accepted I'll review the entire flow, which includes building the electron app and installing it on my computer. Once that's done I believe the pop-up will say "Open New Expensify".

**Comment by michaelhaxhiu** [comment]
This is a `New Feature`, but @cowboycito has a valid proposal that we are considering implementing even though it's a feature freeze. For the next BZ assigned, please keep an eye and let's make sure we either guide it forward or close depending on the outcome.

**Comment by NikkiWines** [comment]
@Santhosh-Sellavel any further thoughts on this proposal? Seems like @cowboycito may have addressed your last comment

**Comment by Santhosh-Sellavel** [comment]
This is what I see 😕 ,

I cloned your code 
```
 npm i
 // Run both web & desktop from different terminal instances
 yarn desktop
 yarn web
```
https://user-images.githubusercontent.com/85645967/202848587-9bf8e5e3-b6af-4504-93c9-1e231784ea35.mov

cc: @cowboycito Can you help me?

@NikkiWines Have you tested if it works, we can loop in the design team for the intermediate page design!

**Comment by cowboycito** [comment]
Given how #10614 turned out I have no interest in continuing working on this. If someone else wants to get my proposal and elaborate on it feel free to do so, but keep in mind that this is a feature request and the issue can get closed at any moment without further notice.

**Comment by Christinadobrzyn** [comment]
Hey @cowboycito, sorry to hear about how things went on the other project, is the issue that you spent a lot of time working on a proposal and it was closed without compensation? If that's the case, I'm happy to reach out to our team about this for you. 

In terms of this project, I'm not sure what the next steps are, should we continue to work on this, put it on [Hold] until something else is finished or close and discuss compensation for what has been done?

**Comment by cowboycito** [comment]
@Christinadobrzyn that's exactly what happened, and honestly I was impressed with the lack of respect on the way the issue was closed, with total disregard for the time I spent working on it. Given that #10614 and this both issues are feature requests, not bugs, I don't wanna risk wasting my time again working on something that can get closed at any second.

**Comment by Santhosh-Sellavel** [comment]
@cowboycito close to completing the solution there, I've started a discussion with hax for the discussing next steps consider reopening if possible but hax is OOO. So initiated a thread [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1669126480589029)

**Comment by Santhosh-Sellavel** [comment]
bump @cowboycito

**Comment by cowboycito** [comment]
@Santhosh-Sellavel I did a bunch of tests and I was not able to replicate the issue you're having. I wonder if you have Electron installed as an app on your computer. If that's the case, try uninstalling it and giving it another shot.

**Comment by Santhosh-Sellavel** [comment]
>  I wonder if you have Electron installed as an app on your computer.

Yeah may be I did, I test on a different pc it works.
@NikkiWines Looks good to me, can we loop in design team for intermediate page design thanks!

**Comment by Santhosh-Sellavel** [comment]
Waiting for the intermediate page design!

**Comment by JmillsExpensify** [comment]
@NikkiWines @Christinadobrzyn @Santhosh-Sellavel Can I get some context where we decided to work on this feature request? I might have missed it, but I think it's critically important that we focus everyone in our open source ecosystem on bugs right now. It unfortunately means that we need to pause all feature development as a result – as the core of the platform is still shaky and will be until we get to the point where all bugs are solved within a month of being logged. Again, I'd love to get back to feature development, though achieving stability requires hard decisions just like this one.

**Comment by Santhosh-Sellavel** [comment]
@JmillsExpensify Check this https://github.com/Expensify/App/issues/10893#issuecomment-1320409207

**Comment by shawnborton** [comment]
Here is a design idea:
![image](https://user-images.githubusercontent.com/2319350/204376163-16c887c9-aa8f-45e2-a008-d6cd2fbd74f9.png)

**Comment by NikkiWines** [comment]
Love it @shawnborton that looks so cool. This feature would also be useful in the ongoing #vip-passwordless project for deeplinking directly to newDot desktop.

**Comment by Christinadobrzyn** [comment]
[Asking](https://expensify.slack.com/archives/C02NK2DQWUX/p1669691942868689) the team and members of the vip-passwordless project to give their thoughts on moving forward with this job.

**Comment by JmillsExpensify** [comment]
> This feature would also be useful in the ongoing #vip-passwordless project for deeplinking directly to newDot desktop.

@NikkiWines Ok I think that context as well as what @Santhosh-Sellavel posted are fair reasons to keep this moving forward. Thanks!

**Comment by cowboycito** [comment]
@Santhosh-Sellavel @NikkiWines @JmillsExpensify any idea on when we'll have the designs developed and approved so I can start working on implementing it?

**Comment by shawnborton** [comment]
I'd say we're good to go with the design I posted above.

**Comment by cowboycito** [comment]
@shawnborton is there any way I can have access to the design spec so I'm able to get references for colors, fonts, icons...?

**Comment by shawnborton** [comment]
Figma is here: https://www.figma.com/file/WlrZyRCTQ0LFoKcjSmtTsd/General-UI-Patterns?node-id=1634%3A14718&t=JpwlqDQC4vNBvr3N-1

SVG to use is here: [empty-state__launchdesktop.svg.zip](https://github.com/Expensify/App/files/10118056/empty-state__launchdesktop.svg.zip)

Note that the headline font hasn't been added to the repo yet, that is being worked on in a separate issue. In the meantime you can just use the normal body font family for now.

**Comment by cowboycito** [comment]
I don't have access to that Figma file

**Comment by shawnborton** [comment]
Sorry, try again!

**Comment by cowboycito** [comment]
Thank you!

**Comment by Christinadobrzyn** [comment]
Sorry, I'm losing track of this one - can you please let me know if we've concluded who to hire for this yet or are we still reviewing proposals? @Santhosh-Sellavel @NikkiWines

**Comment by Santhosh-Sellavel** [comment]
@Christinadobrzyn we almost concluded to hire @cowboycito. Since @cowboycito new contributor I'm waiting for a complete proposal. I think we can get it today, @cowboycito  can you send an updated proposal with design changes, and quickly review thanks!

cc: @NikkiWines

**Comment by cowboycito** [comment]
@Santhosh-Sellavel I have just pushed a commit to the `iss-10893/desktop-deeplink` branch implementing the design provided by @shawnborton. This is how it looks like:

<img width="1624" alt="Screen Shot 2022-11-30 at 20 49 41" src="https://user-images.githubusercontent.com/47456024/204932589-53a37c46-5cb3-41e6-ad77-bae65d20c74d.png">

**Comment by Santhosh-Sellavel** [comment]
@NikkiWines 

@cowboycito's proposal looks great to me, lets's hire them.
C+ Reviewed
🎀 👀 🎀 

@cowboycito Few things as this are your first PR

Please ensure you read the PR Guidelines and fill out the Author checklists carefully.
Don't add/remove anything irrelevant.

@NikkiWines 
Please helps us get you us Spanish copy of the new messages. With that we are good to kickstart the PR process, thanks, everyone!

cc: @Christinadobrzyn

**Comment by Santhosh-Sellavel** [comment]
@cowboycito One bug I noticed while briefly testing, the intermediate page is skipped in safari

https://user-images.githubusercontent.com/85645967/204962490-c66541c6-e160-4621-bf26-1962e243d938.mov

**Comment by Christinadobrzyn** [comment]
Old job is closed - created a new one - https://www.upwork.com/jobs/~015b67b5fe4aef3990

@NikkiWines let me know if all of this https://github.com/Expensify/App/issues/10893#issuecomment-1333109497 looks good!

**Comment by cowboycito** [comment]
@Santhosh-Sellavel I'll do some more tests on multiple different browsers later today

**Comment by NikkiWines** [comment]
@Christinadobrzyn yep looks good! Please hire @cowboycito for this job! 🙌

**Comment by NikkiWines** [comment]
> @NikkiWines
Please helps us get you us Spanish copy of the new messages. With that we are good to kickstart the PR process, thanks, everyone!

Yep! Once we have a PR up I'll help get us translations!

**Comment by Christinadobrzyn** [comment]
Hired in Upwork: 

- @Santhosh-Sellavel $2000 for C+
- @cowboycito - will you please send your Upwork profile so I can hire you? Thanks!

**Comment by cowboycito** [comment]
Hey @Christinadobrzyn ! My Upwork profile is https://www.upwork.com/freelancers/~0110213b536b398fa9

**Comment by cowboycito** [comment]
Just an update here as I'm still investigating the deeplink issue in Safari.
I accidentally found out that if you continuously keep moving our mouse along the screen the deeplink will work in Safari (the "Launching Expensify" screen will be shown). I still don't know exactly why this happens, but I feel like this is on purpose since Safari will only fire the window events we care about if it "detects" there's a user action ongoing (like when you're moving your mouse). The biggest problem is that I can't find pretty much anything online about how other companies are handling that, so what I'm doing is trying to take a look at Slack's source code to understand how they handle this issue so smoothly. One thing I was able to notice is that if you open Slack's deeplink URL in Safari (and don't move your mouse) you can see that the loading bar (just under the URL bar) keeps showing like it's loading something, but once you move your mouse just a bit, the loading bar finishes the animation and then the window events are successfully fired. That's one approach I wanna try.

cc: @Santhosh-Sellavel @NikkiWines

**Comment by Santhosh-Sellavel** [comment]
@cowboycito Any further updates on this?

**Comment by NikkiWines** [comment]
Not overdue, currently being worked on! @cowboycito any updates at the moment?

**Comment by Santhosh-Sellavel** [comment]
@cowboycito Any update?

**Comment by cowboycito** [comment]
@Santhosh-Sellavel not yet unfortunately. This issue in Safari is really tricky and I was not able to find a solution yet, but I keep trying.

**Comment by NikkiWines** [comment]
Hmm, @cowboycito if you want to push up a PR for now @Santhosh-Sellavel and I can begin reviewing across the other platforms. 

> One thing I was able to notice is that if you open Slack's deeplink URL in Safari (and don't move your mouse) you can see that the loading bar (just under the URL bar) keeps showing like it's loading something, but once you move your mouse just a bit, the loading bar finishes the animation and then the window events are successfully fired. That's one approach I wanna try.

I wasn't able to reliably reproduce this (for slack links) after granting safari the ability to open slack links. But, before I had done that, was able to reproduce the behavior you mentioned with loading indicator halting until mouse movement occurs. 

https://user-images.githubusercontent.com/16747822/206043140-ce94f34b-dd5b-4e91-9fda-c7c8cdd9e949.mov

@cowboycito If your branch has your latest chanes I'm happy to test to see if the same behavior occurs for me.

**Comment by cowboycito** [comment]
@NikkiWines @Santhosh-Sellavel  I feel like I'm really getting close to the exact behavior implemented by Slack and Zoom. Give me until tomorrow if possible and I think I'll get a PR done with this working on all browsers. If that's not the case we can discuss further steps.

**Comment by cowboycito** [comment]
@NikkiWines @Santhosh-Sellavel I did lots of digging and unfortunately I was not able to make it work as I thought I could.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @cowboycito, let us if there's any update!

@NikkiWines What can we do now?
1. Let's wait for a new proposal to address everything
2. Continue with @cowboycito proposal. Track the Safari problem in this issue or a new issue, once the current job is complete.

cc: @Christinadobrzyn

**Comment by Santhosh-Sellavel** [comment]
Bump @NikkiWines

**Comment by NikkiWines** [comment]
Thanks for the bump @Santhosh-Sellavel, missed this last update. 

To be clear, @cowboycito, what is the current behavior with safari on your branch? Is it just that you have to move your mouse in order for the deep link to initiate? How much does it differ from the behavior of slack or spotify links on the same browser? 

Once I have a better understanding of that, I think we can decide on whether to go with option 1 or 2 that @Santhosh-Sellavel outlined above. Ideally, I think option 2 would be best, just to so that we can retain the work that's already been done, rather than having to start from scratch.

**Comment by cowboycito** [comment]
@NikkiWines deeplink works on all browsers. The problem is actually checking if the deeplink "worked" (the user has the app installed on their computer) or not. On chrome that's doable and it's working fine, but in Safari it only works if you "do something" (like move your mouse or press any random key in the keyboard in the given timeout we have). So the current behavior in chrome is correct: if the user has the app installed it will show the `Launching Expensify` screen, and if the app is not installed it will go straight to that page using the browser. The same happens in Safari if the user moves the mouse, but if the mouse is not moved (which will cause for the event we're interested at to not get fired) the deeplink popup will show as expected, but we're not gonna see the `Launching Expensify` screen.
I did some really good digging on Slack and Zoom's source code (although it was minified so it was really hard to understand everything that was going on) and I'm pretty sure my implementation is really similar to theirs. One thing I noticed is that none of those apps have the exact same experience of what we're looking for (show a "launching" screen if the desktop app is installed or show the app in the browser if the desktop app is not installed). From what I've tested so far, if you open a Slack message link on a computer that doesn't have the Slack desktop app installed it will not take the user straight to the Slack browser app (at least not in Safari), it will always show the "launching" screen with the "open in browser" button. And Zoom doesn't have a web app, so they don't have to worry about that.
About Spotify (which I actually didn't notice was an example I could take a look at) opens the deeplink popup and at the same time loads the web app behind it, as in this image:

<img width="1719" alt="Screen Shot 2022-12-09 at 21 29 56" src="https://user-images.githubusercontent.com/47456024/206818768-56a92ac2-6cf8-43c1-9b97-a9590476bd82.png">

That would pretty much be the behavior on Expensify on Safari: show the deeplink popup but also load the page as usual.

I don't wanna say that it's impossible to do what we want in Safari, but if it is I think no one found out yet.

cc: @Santhosh-Sellavel

**Comment by cowboycito** [comment]
@Santhosh-Sellavel I already did all the code changes I wanted to do and pushed it to my branch. Will do the PR guidelines workflow tomorrow morning and open up the PR so you can review the code.

**Comment by Santhosh-Sellavel** [comment]
Sounds good!

**Comment by cowboycito** [comment]
@Santhosh-Sellavel [PR](https://github.com/Expensify/App/pull/13606) created

**Comment by Christinadobrzyn** [comment]
I'll be ooo till Jan 9th, assigning this issue to @mallenexpensify. Matt, I'm happy to take it back when I return!

**Comment by Santhosh-Sellavel** [comment]
Note all this PR is gone past 7+ biz days counting from this [date](https://github.com/Expensify/App/issues/10893#issuecomment-1346633089) as we decide to move forward to PR here ignoring the safari case for now.

And I believe this might take some more time I guess, especially since this is a new feature what we are running into is a completely new learning for us!

Requesting here to consider flagging this one from the penalty move to it a flat job fee. Also, I might request a compensation increase on this one, but since the job is not over it's too early not sure how much work is left I will keep that discussion for a later date, thanks!

cc: @NikkiWines @mallenexpensify  @Christinadobrzyn

**Comment by cowboycito** [comment]
@Santhosh-Sellavel what is a penalty fee and how does that affect me?

**Comment by Santhosh-Sellavel** [comment]
check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md)

**Comment by cowboycito** [comment]
I think we should discuss more about that before I continue working on that PR

**Comment by JmillsExpensify** [comment]
Please continue working on the PR. This is a feature request, I fully expect the PR to take longer given that we don't have a formal design in place (though note for internal colleagues: this is why we need to be very careful scoping features in GH, there's a high chance of unforeseen gotchas to come up, which is mostly a reflection of us not planning well internally and have crisp requirements imo).

**Comment by mallenexpensify** [comment]
@NikkiWines , when you're back, can you review the PR plzzz https://github.com/Expensify/App/pull/13606

**Comment by Santhosh-Sellavel** [comment]
PR is merged, awaiting deployment!

**Comment by Christinadobrzyn** [comment]
I'll take this GH back @mallenexpensify thank you for watching it!

**Comment by Santhosh-Sellavel** [comment]
@Christinadobrzyn 

We can continue the discussion regarding the bounty increase [request](https://github.com/Expensify/App/issues/10893#issuecomment-1305900616) on this one while we wait the regression period.

cc: @cowboycito

**Comment by cowboycito** [comment]
@Santhosh-Sellavel how should we proceed for payment?
cc @Christinadobrzyn @NikkiWines

**Comment by Santhosh-Sellavel** [comment]
You've requested an increase in compensation while working on the proposal [here](https://github.com/Expensify/App/issues/10893#issuecomment-1305900616). 

But since we are not sure about how long/complex it might take, I've replied [here](https://github.com/Expensify/App/issues/10893#issuecomment-1305911115)
> Again this is too early, we can discuss it once the issue is completed.

Now the job is done I've initiated the discussion [here](https://github.com/Expensify/App/issues/10893#issuecomment-1383274376) waiting for it CM. 

Meanwhile, let us know what you think is a fair amount. I'll post internally let us get it sorted thanks!

**Comment by mallenexpensify** [comment]
@Santhosh-Sellavel do you want to post the increased compensation request first then @cowboycito can 👍 or comment? I think that'd be _easiest_.  Please include the amount total you think is fair. Provide reasoning, including links and details. That will make it easy for the BZ / @Christinadobrzyn  to 👍 (if they agree).  Thanks.

**Comment by Christinadobrzyn** [comment]
Hey @cowboycito - what compensation amount do you think is fair for this job?

**Comment by Santhosh-Sellavel** [comment]
Sorry for the delay everyone, I'll post a summary of events by date.
The current price of the job of $2000 is set on 22 Oct 2022 [here](https://github.com/Expensify/App/issues/10893#issuecomment-1287352364) to encourage proposals.

Pretty same we got the proposal kick-started from @cowboycito [here](https://github.com/Expensify/App/issues/10893#issuecomment-1287397566)

After review and some discussions, I've requested additional changes i.e requested POC to completely test the code on  7 Nov 2022 [here](https://github.com/Expensify/App/issues/10893#issuecomment-1305000930).

On the same day, @cowboycito questioned about an increase in bounty [here](https://github.com/Expensify/App/issues/10893#issuecomment-1305900616).

I thought it's fair to raise bounty since I'm not sure about how much work is left out, so commented [here](https://github.com/Expensify/App/issues/10893#issuecomment-1305911115) let's discuss once the job is complete.

On 10 Nov 2022 [here](https://github.com/Expensify/App/issues/10893#issuecomment-1309544197) the POC is ready for review.

After some discussions and review. On 25 Nov 2022 [here](https://github.com/Expensify/App/issues/10893#issuecomment-1327692738), I was able to verify the POC.  And a design request for a deep link page was initiated.

The proposal was updated by @cowboycito on Dec 1 [here](https://github.com/Expensify/App/issues/10893#issuecomment-1332906259) And I recommended hire [here](https://github.com/Expensify/App/issues/10893#issuecomment-1333109497)  on the same day.

Also, I noticed a [bug](https://github.com/Expensify/App/issues/10893#issuecomment-1333134115) on safari while reviewing. 

9 Dec 2022
@cowboycito Investigated it but could not get through so requested what's next steps [here](https://github.com/Expensify/App/issues/10893#issuecomment-1343254373)

12 Dec 2022 Ignoring the bug, we have a lot to ship here so we decided to move forward [here](https://github.com/Expensify/App/issues/10893#issuecomment-1344936046) and [here](https://github.com/Expensify/App/issues/10893#issuecomment-1346633089)

15 Dec 2022 [PR](https://github.com/Expensify/App/pull/13606#issue-1497617051) Submitted

23 Dec 2022 [Here](https://github.com/Expensify/App/issues/10893#issuecomment-1364150071) I requested to flag this one as a penalty as its feature might take more time.

25 Dec 2022 to Jan 2 No activity due to Holiday. 

Resumed Review Activity Jan 3 PR was [merged](https://github.com/Expensify/App/pull/13606#issuecomment-1376347412) on Jan 10.

Considering the efforts made here, IMO $5K is a fair amount here.

**Comment by cowboycito** [comment]
Sorry for the delay everyone! I agree with what @Santhosh-Sellavel said and I think that sounds fair.

**Comment by Christinadobrzyn** [comment]
Thanks @Santhosh-Sellavel and @cowboycito! @NikkiWines does this sound good to you?

**Comment by iwiznia** [comment]
Context https://expensify.slack.com/archives/C01GTK53T8Q/p1674481724781139

I think this PR has a bug: whenever I reload the website, I get the `Open electron` banner, but I don't have the app installed (I think, not 100% how to check).  So, it's kind of 2 bugs I think?
1. Banner should not be shown if you are reloading the page
2. Banner should not be shown if you don't have the app installed

**Comment by NikkiWines** [comment]
> Thanks @Santhosh-Sellavel and @cowboycito! @NikkiWines does this sound good to you?

Yep, seems reasonable to me! This was a tricky issue and both @cowboycito and @Santhosh-Sellavel did a good job of keeping it moving and following through to get it across the finish line 👍

**Comment by Christinadobrzyn** [comment]
Okay! ready to close this out: 

- Paid @cowboycito $2,000 for the job and $3,000 bonus
- Paid @Santhosh-Sellavel $2,000 for the job and $3,000 bonus
- No external reporter
- Closed the Upwork job
- Closing this GH as complete! 

please let me know if there are any questions!
